src/edu/csus/ecs/pc2/core/list/UnavailableRunsList.java

```diff
@@ -0,0 +1,202 @@
+     * have been added to the list previously are given increasingly longer (further in the future) expiration times.
+     * 
+     * Note that runs are not "automatically" removed from the list when their "expiration time" has arrived; a call to
+     * method {@link #removeExpiredRuns()} must be made to cause expired runs to be removed
+     * (see method {@link AutoJudgingMonitor#findNextAutoJudgeRun()}). 
+     * 
+     * @param run the Run to be added to the list.
+     */
+    public void addRun(Run run) {
+        
+        if (contest==null) {
+            throw new RuntimeException("UnavailableRunsList contains null contest");
+        }
+        
+        if (run==null) {
+            if (log!=null) {
+                log.warning("Attempted to add a null run to the UnavailableRunsList!");
+            } else {
+                System.err.println("Attempted to add a null run to the UnavailableRunsList, and no log is available for logging!!");
+                throw new RuntimeException("Attempted to add a null run to the UnavailableRunsList, and no log is available for logging!!");
+            }
+        }
+        
+        //find the number of times the run has previously been added to the UnavailableRunsList (if any).
+        //  (Note that we're checking the hashtable of ALL runs which have EVER been added to the UnavailableRunsList, 
+        //  not the current UnavailableRunsList.) The "allAddedRuns" hashtable maps Runs (via their ElementId) to a count 
+        //  of the number of times that run has ever been added to the UnavailableRunsList.
+        int timesAddedToRunList;
+        if (allAddedRuns.containsKey(run.getElementId())) {
+            
+            //the run has previously been added; get the count of number of times it has been added previously
+            timesAddedToRunList = allAddedRuns.get(run.getElementId());
+            
+            //indicate another addition is occurring
+            timesAddedToRunList++ ;
+            allAddedRuns.put(run.getElementId(), timesAddedToRunList);
+            
+        } else {
+            
+            //this is the first time the run has been added
+            allAddedRuns.put(run.getElementId(),  1);
+            timesAddedToRunList = 1;
+        }
+        
+        //compute how long past "now", as a function of number of times already added, the run should "expire" on the UnavailableRunsList.
+        // (Note that the current calculation is a LINEAR increase in time; it would probably be better to have something like an EXPONENTIAL
+        //   increase.  However, since this entire class is (hopefully) temporary, I didn't spend any more time on that... jlc)
+        long contestTimeNowInSecs = contest.getContestTime().getElapsedSecs();
+        long expirationTimeInSecs = contestTimeNowInSecs + (INITIAL_EXPIRATION_TIME_OFFSET_SECS * timesAddedToRunList) ;
```

Comment:
Since this is temporary, this is more of a comment than a correction.
Could adding the timesAddedToRunList * INITIAL_EXPIRATION_TIME_OFFSET_SECS  cause a run to effectively never expire if the run not available occurs multiple times, like the condition that prompted this issue?
Since this is temporary, I will not suggest removing the timesAddedToRunList  entirely, nor creating
an upper limit for timesAddedToRunList .

Reply:
In answer to the question, `Could adding the timesAddedToRunList * INITIAL_EXPIRATION_TIME_OFFSET_SECS cause a run to effectively never expire...`:  I don't think so.  I say this because the only way the "expiration time" is ever added is when the run gets put on the "unavailableRuns list".  Once it is on that list, the AJ will never request that run again until its expiration time is reached (that's the whole purpose of the UnavailableRuns list -- to stop the AJ from requesting that run).  Since the AJ will never request that run until it comes off the list, it will never receive that run again until after it "expires" and comes off the list.

Having said that, can I invent a scenario where the AJ *could* receive that run again?  Sure:  the AJ could request run 1, get back "unavailable", put it on the "UnavailableList", then request run 2; when the server gets the request for run 2 it could instead send back run 1 in place of run 2.  But if that kind of thing is happening we've got way more problems than the issue that this PR is addressing.

Given that you've said this is a comment and not a requested fix, I'm going to mark this "resolved".  If anyone wants to spend time programming for the outlandish situation I just described (for some code that's -- hopefully -- temporary, please feel free to file a separate Issue and Go For It.   I don't think it belongs any more as part of this Issue/PR.